core/keychain/keychain.go

```diff
@@ -2,40 +2,98 @@ package keychain
 
 import (
 	"crypto/ed25519"
-	"crypto/sha1"
-
-	"golang.org/x/crypto/pbkdf2"
+	"encoding/hex"
+	"strings"
 
 	"errors"
 
-	db "github.com/FleekHQ/space-daemon/core/store"
+	"github.com/99designs/keyring"
+	ri "github.com/FleekHQ/space-daemon/core/keychain/keyring"
+	"github.com/FleekHQ/space-daemon/core/store"
 	"github.com/libp2p/go-libp2p-core/crypto"
 )
 
-const PrivateKeyStoreKey = "private_key"
-const PublicKeyStoreKey = "public_key"
+const PrivateKeyStoreKey = "key"
+const PublicKeyStoreKey = "pub"
+
+const privKeyMnemonicSeparator = "___"
 
 var (
 	ErrKeyPairNotFound = errors.New("No key pair found in the local db.")
 )
 
 type keychain struct {
-	store db.Store
+	fileDir string
+	st      store.Store
+	ring    ri.Keyring
 }
 
 type Keychain interface {
 	GenerateKeyPair() (pub []byte, priv []byte, err error)
 	GenerateKeyFromMnemonic(...GenerateKeyFromMnemonicOpts) (mnemonic string, err error)
 	GetStoredKeyPairInLibP2PFormat() (crypto.PrivKey, crypto.PubKey, error)
+	GetStoredPublicKey() (crypto.PubKey, error)
 	GenerateKeyPairWithForce() (pub []byte, priv []byte, err error)
-	GeneratePasswordBasedKey(password string) (key, salt []byte, iterations int)
 	Sign([]byte) ([]byte, error)
 	ImportExistingKeyPair(priv crypto.PrivKey) error
 }
 
-func New(store db.Store) *keychain {
+type keychainOptions struct {
+	fileDir string
+	store   store.Store
+
+	// Don't use kc.ring directly, use getKeyRing() instead
+	ring ri.Keyring
+}
+
+var defaultKeychainOptions = keychainOptions{
+	fileDir: store.DefaultRootDir,
+}
+
+// Helper function for setting keychain file path for Windows/Linux
+func WithPath(path string) Option {
+	return func(o *keychainOptions) {
+		if path != "" {
+			o.fileDir = path
+		}
+	}
+}
+
+func WithStore(st store.Store) Option {
+	return func(o *keychainOptions) {
+		if st != nil {
+			o.store = st
+		}
+	}
+}
+
+// Used to inject a mock keyring in tests or in case you want to use a custom keyring implementation
+func WithKeyring(ring ri.Keyring) Option {
+	return func(o *keychainOptions) {
+		if ring != nil {
+			o.ring = ring
+		}
+	}
+}
+
+type Option func(o *keychainOptions)
+
+func New(opts ...Option) *keychain {
+	o := defaultKeychainOptions
+	for _, opt := range opts {
+		opt(&o)
+	}
+
+	defaultStore := store.New(store.WithPath(o.fileDir))
```

Comment:
Should we move this into the `if` below so we don't need to create it if it was already supplied?

Reply:
Fixed in 8d0b7d5